src/Common/OpacityTable.cpp

```diff
@@ -0,0 +1,308 @@
+//=================================================================================================
+template <int ndim>
+OpacityTable<ndim>::OpacityTable
+ (string radws_table,
+  int lombardi_,
+  SimUnits *simunits)
+{
+  // Set lombardi method flag
+  lombardi = lombardi_;
+
+  int i, j, l;
+  DOUBLE eos_dens_, eos_temp_, eos_energy_, eos_mu_, kappa_, kappar_, kappap_, eos_gamma_;
+  DOUBLE num, denom, tempunit, fcol;
+  string line;
+
+  // Check for correct opacity units
+  if (simunits->kappa.outunit != "cm2_g") {
+    ExceptionHandler::getIstance().raise("Error: Wrong units for opacity, use cm2_g");
+  }
+
+  ifstream file;
+  file.open(radws_table.c_str(), ios::in);
+  if (!file.is_open()) {
+    ExceptionHandler::getIstance().raise("Error: Cannot open opacity table!");
+  }
+
+  //-----------------------------------------------------------------------------------------------
+  if (file.good()) {
+    do {
+      getline(file, line);
+    } while (line[0] == '#');
+    istringstream istr(line);
+    istr >> ndens >> ntemp >> fcol;
+
+    eos_dens     = new FLOAT[ndens];
+    eos_temp     = new FLOAT[ntemp];
+    eos_energy   = new FLOAT*[ndens];
+    eos_mu       = new FLOAT*[ndens];
+    eos_gamma    = new FLOAT*[ndens];
+    kappa_table  = new FLOAT*[ndens];
+    kappar_table = new FLOAT*[ndens];
+    kappap_table = new FLOAT*[ndens];
+
+    for (i = 0; i < ndens; ++i){
+      eos_energy[i]   = new FLOAT[ntemp];
+      eos_mu[i]       = new FLOAT[ntemp];
+      eos_gamma[i]    = new FLOAT[ntemp];
+      kappa_table[i]  = new FLOAT[ntemp];
+      kappar_table[i] = new FLOAT[ntemp];
+      kappap_table[i] = new FLOAT[ntemp];
+    }
+
+    // read table
+    i = 0;
+    l = 0;
+    j = 0;
+
+    //---------------------------------------------------------------------------------------------
+    while (getline(file, line)) {
+      istringstream istr(line);
+
+      if (istr >> eos_dens_ >> eos_temp_ >> eos_energy_ >> eos_mu_>> kappa_ >> kappar_ >> kappap_ >> eos_gamma_) {
+
+        eos_energy[i][j]   = eos_energy_/(simunits->u.outscale * simunits-> u.outcgs);
+        eos_mu[i][j]       = eos_mu_;
+        eos_gamma[i][j]    = eos_gamma_;
+        kappa_table[i][j]  = kappa_/(simunits->kappa.outscale * simunits-> kappa.outcgs);
+        kappar_table[i][j] = kappar_/(simunits->kappa.outscale * simunits-> kappa.outcgs);
+        kappap_table[i][j] = kappap_/(simunits->kappa.outscale * simunits-> kappa.outcgs);
+
+        if (l < ntemp) {
+          eos_temp[l] = log10(eos_temp_/(simunits->temp.outscale * simunits->temp.outcgs));
+        }
+
+        ++l;
+        ++j;
+
+        if (!(l % ntemp)) {
+          eos_dens[i] = log10(eos_dens_/(simunits->rho.outscale * simunits-> rho.outcgs));
+          ++i;
+          j = 0;
+        }
+
+      }
+      else {
+        ExceptionHandler::getIstance().raise("Error: The specified opacity file cannot be read!");
+      }
+
+    }
+    file.close();
+  }
+
+  // Set fcol2
```

Comment:
I think fcol should be associated with the energy integration object rather than the Opacity table. I don't see why the Opacity cares about how it is being used.